python/ray/data/tests/test_dataset.py

```diff
@@ -142,6 +142,19 @@ def __call__(self, x):
     assert len(actor_reuse) == 10, actor_reuse
 
 
+def test_transform_failure(shutdown_only):
+    ray.init(num_cpus=2)
+    ds = ray.data.from_items([0, 10], parallelism=2)
+
+    def mapper(x):
+        time.sleep(x)
+        assert False
+        return x
+
+    with pytest.raises(ray.exceptions.RayTaskError):
```

Comment:
What did it use to throw before this PR?

Reply:
It threw a `RayTaskError` before as well! Since we're now catching these task errors and doing cleanup, this test is here to guard against a regression where we don't correctly reraise the exception.